File: codequest/pathfinder.py
def fill(maze,mask,x,y):
	if x==len(maze)-1 and y==len(maze)-1:
		return 1
	
	ret = 0
	mask[x][y] = True
	ln = len(maze)-1
	
	if x>0 and not mask[x-1][y]:
		ret += fill(maze,mask,x-1,y)
	
	if y>0 and not mask[y][x-1]:
		ret += fill(maze,mask,x,y-1)
	
	if x<ln and not mask[x+1][y]:
		ret += fill(maze,mask,x+1,y)
	
	if y<ln and not mask[x][y+1]:
		ret += fill(maze,mask,x,y+1)
	
	mask[x][y] = False
	return ret

def test_case():
	n = int(input())
	
	maze = []
	mask = []
	
	for i in range(n):
		maze.append([i=='1' for i in input()])
		mask.append([False for i in range(n)])
	
	ret = fill(maze,mask,0,0)

t = int(input())

for i in range(t):
	test_case()

File: contests/codequest/pathfinder.cpp
#include <iostream>
using namespace std;

long long fill(bool map[],bool mask[],long long n,long long x,long long y){
	if(x==n-1&&y==n-1)return 1;
	long long ret=0;
	mask[x*n+y]=1;
	if(x>0&&!mask[x*n-n+y]&&map[x*n-n+y])
		ret+=fill(map,mask,n,x-1,y);
	if(y>0&&!mask[x*n+y-1]&&map[x*n+y-1])
		ret+=fill(map,mask,n,x,y-1);
	if(x<n-1&&!mask[x*n+n+y]&&map[x*n+n+y])
		ret+=fill(map,mask,n,x+1,y);
	if(y<n-1&&!mask[x*n+y+1]&&map[x*n+y+1])
		ret+=fill(map,mask,n,x,y+1);
	mask[x*n+y]=0;
	return ret;
}

void test_case(){
	long long n;cin>>n;
	bool map[n*n],mask[n*n];
	for(long long i=0;i<n;++i)
		for(long long j=0;j<n;++j){
			char x;cin>>x;
			map[i*n+j]=(x=='1');
			mask[i*n+j]=0;
		}
	long long ret=fill(map,mask,n,0,0);
	cout<<ret<<'\n';
}

int main(){
	int t;cin>>t;
	while(t-->0)test_case();
}
